Part 1: Data Loading and Exploration (20 points)

In [245]:
# Load the California Housing dataset from sklearn.datasets.

from sklearn.datasets import fetch_california_housing

# Load the California housing dataset
housing = fetch_california_housing()

In [246]:
# Create a DataFrame for the features and a Series for the target variable (med_house_value). 
import pandas as pd

X = pd.DataFrame(housing.data, columns=housing.feature_names) 
y = pd.Series(housing.target, name='med_house_value')

In [247]:
# Perform an initial exploration of the dataset:
# Display the first five rows of the dataset. (5 pts)
print("Dataset preview:")
X.head()

Dataset preview:


,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25


In [248]:
# Print the feature names and check for missing values. (5 pts)
print("Feature names: ", X.columns.tolist())

print("\nCheck for missing values:")
X.isnull().sum()

Feature names:  ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude']

Check for missing values:


MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
dtype: int64

In [249]:
# Generate summary statistics (mean, min, max, etc.). (10 pts)
print("Summary statistics:")
print(X.describe())

Summary statistics:
             MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  20640.000000  20640.000000  20640.000000  20640.000000  20640.000000   
mean       3.870671     28.639486      5.429000      1.096675   1425.476744   
std        1.899822     12.585558      2.474173      0.473911   1132.462122   
min        0.499900      1.000000      0.846154      0.333333      3.000000   
25%        2.563400     18.000000      4.440716      1.006079    787.000000   
50%        3.534800     29.000000      5.229129      1.048780   1166.000000   
75%        4.743250     37.000000      6.052381      1.099526   1725.000000   
max       15.000100     52.000000    141.909091     34.066667  35682.000000   

           AveOccup      Latitude     Longitude  
count  20640.000000  20640.000000  20640.000000  
mean       3.070655     35.631861   -119.569704  
std       10.386050      2.135952      2.003532  
min        0.692308     32.540000   -124.350000  
25%        2.42974

Part 2: Linear Regression on Unscaled Data (30 points)

In [250]:
# Split the dataset into training and test sets (80% training, 20% testing). (5 pts)
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train a linear regression model on the unscaled data using sklearn.linear_model.LinearRegression. (5 pts)
from sklearn.linear_model import LinearRegression

lin_reg_raw = LinearRegression()
lin_reg_raw.fit(X_train, y_train)

# Make predictions on the test set. (5 pts)
y_pred = lin_reg_raw.predict(X_test)

In [251]:
# Evaluate model performance using the following metrics:
from sklearn.metrics import mean_squared_error, root_mean_squared_error, r2_score
print("Unscaled data model:")

# Mean Squared Error (MSE) (5 pts)
mse_raw = mean_squared_error(y_test, y_pred)
print(f"Mean squared error: {mse_raw:.2f}")

# Root Mean Squared Error (RMSE) (5 pts)
rmse_raw = root_mean_squared_error(y_test, y_pred)
print(f"Root squared error: {rmse_raw:.2f}")

# R² Score (5 pts)
r2_raw = r2_score(y_test, y_pred)
print(f"R² score: {r2_raw:.2f}")

Unscaled data model:
Mean squared error: 0.56
Root squared error: 0.75
R² score: 0.58


In [252]:
# View model's coefficients
print("Model coefficients (unscaled):")
print(pd.Series(lin_reg_raw.coef_, index=X.columns))

Model coefficients (unscaled):
MedInc        0.448675
HouseAge      0.009724
AveRooms     -0.123323
AveBedrms     0.783145
Population   -0.000002
AveOccup     -0.003526
Latitude     -0.419792
Longitude    -0.433708
dtype: float64


Interpretation Questions:

What does the R² score tell us about model performance?
- The R² score is a measure of the proportion of variance in the target variable that is explained by the predictor variables. A score of 1 indicates that the model is a perfect fit, and 100% of the variation in the target can be explained by model. A score of 0 indicates that the model fails to explain variance in the target variable. This model's R² score is 0.58, meaning it can explain 58% of variance in target variable predictions. This indicates that the model is a moderate fit since just over half of the variation of predicted outcomes can be accounted for by the model.

Which features seem to have the strongest impact on predictions based on the model’s coefficients?
- To determine the relative impact different features have on the model's predictions for med_house_value (median house value in units of 100,000), it is essential to compare model coefficients. Each coefficient reflects the change in performance index per unit change in the respective feature, if all other features were held constant.
- For this model, the three most influential features are AveBedrms, MedInc, and Longitude. AveBdrms has the strongest influence on the target prediction, with a coefficient of 0.783; this means that with each additional bedroom per household, the Performance Index increases by 0.783. MedInc has the second strongest influence on the prediction, with a coefficient of 0.449; this means for every $10,000 increase in median income for houses within a block, the Performance Index increases by 0.449. Finally, Longitude has the third strongest influence on the prediction, with a coefficient of -0.434; this means for every 1 degree increase in longitude, the Performance Index decreases by 0.434.

How well do the predicted values match the actual values?
- Examining the R² Score, MSE, and RMSE together indicates that the predicted values only moderately align with the actual values.
- The R² value of 0.58 indicates that the model explains 58% of variance in the target variable. The closer to 1 the R² value is, the better fit the model, so the score of 0.58 indicates that this model is a moderate fit to the data, as it accounts for over half of the variance. This suggests that predicted values reasonably match the actual values. However, there is room for model improvement since a significant amount of variance remains unexplained.
- The MSE and RMSE indicate that predicted values are a moderate match to actual values. The MSE is 0.56, representing the average squared difference between actual and predicted values; the closer to zero, the better, so a value of 0.56 indicates that this model's predictions are not highly accurate but are still somewhat similar to actual values. The RMSE is the square root of the MSE, which is more easily interpretable in the context of the data since it uses the same units as the target. The RMSE is 0.75, meaning that on average, model predictions deviate from the actual median house value by $75,000 on average. It is important to note that housing prices and the range of costs is extremely large, so even though the RMSE of $75,000 is not excellent, it is also not extremely poor.
- Examined together, the R² score of 0.58, MSE of 0.56, and RMSE of 0.75 indicate that predicted values moderately align with actual values even though they are not a strong match. This suggests that the model has room for improvement but still makes somewhat reasonable estimates based on patterns in the training data.

Part 3: Feature Selection and Simplified Model (25 points)

In [253]:
# Select three features from the dataset to build a simplified model. Explain your choice. (10 pts)
simple_features = ["AveBedrms", "MedInc", "Longitude"]
X_2 = X[simple_features]

X_2.head()

,AveBedrms,MedInc,Longitude
0,1.023810,8.3252,-122.23
1,0.971880,8.3014,-122.22
2,1.073446,7.2574,-122.24
3,1.073059,5.6431,-122.25
4,1.081081,3.8462,-122.25


Explanation of feature choice
- I chose AveBdrms, MedInc, and Longitude to train the simplified model because these three features most strongly influenced the initial model's predictions.
    - AveBdrms had a coefficient of 0.783, indicating that this feature has the greatest influence on median house value predictions. This feature considers how structure and build influences median house value.
    - MedInc had the second largest influence, with a coefficient of 0.449. This feature considers how economic status influences median house value.
    - Longitude has the third strongest influence on predictions and is negatively correlated with the target, with a coefficient of -0.434. This feature considers how geography influences median house value.
- Since these three features are most strongly correlated with the target variable, I believe it is important to include them simplified model. Additionally, the diversity of the three features will allow the simplified model to account for unique architectural, economic, and geographic influences on median house value. Therefore, I believe selecting AveBdrms, MedInc, and Longitude will allow for the most effective and accurate simplified model.

In [254]:
# Train a new linear regression model using only these three features. (5 pts)
X_train_2, X_test_2, y_train_2, y_test_2 = train_test_split(X_2, y, test_size=0.2, random_state=42)

# Train a linear regression model on the unscaled data using sklearn.linear_model.LinearRegression. (5 pts)
lin_reg_raw_2 = LinearRegression()
lin_reg_raw_2.fit(X_train_2, y_train_2)

# Make predictions on the test set. (5 pts)
y_pred_2 = lin_reg_raw_2.predict(X_test_2)

In [255]:
# Evaluate the performance of this simplified model and compare it to the full model. (5 pts)
print("Simplified data model:")

# Mean Squared Error (MSE) (5 pts)
mse_raw_2 = mean_squared_error(y_test_2, y_pred_2)
print(f"Mean squared error: {mse_raw_2:.2f}")

# Root Mean Squared Error (RMSE) (5 pts)
rmse_raw_2 = root_mean_squared_error(y_test_2, y_pred_2)
print(f"Root squared error: {rmse_raw_2:.2f}")

# R² Score (5 pts)
r2_raw_2 = r2_score(y_test_2, y_pred_2)
print(f"R² score: {r2_raw_2:.2f}")

Simplified data model:
Mean squared error: 0.71
Root squared error: 0.84
R² score: 0.46


In [256]:
# View simplified model's coefficients
print("Simplified model coefficients:")
print(pd.Series(lin_reg_raw_2.coef_, index=X_2.columns))

Simplified model coefficients:
AveBedrms   -0.001910
MedInc       0.418949
Longitude   -0.019934
dtype: float64


Interpretation Questions: (5 pts)

How does the simplified model compare to the full model?
- The simplified model performs more poorly than the inclusive model, which is demonstrated by comparing the R² scores, MSE, and RMSE.
    - The simplified model has a lower R² score of 0.46, meaning the model can explain less than half of the varience between predicted and actual values. This R² score is lower than the original model's (0.58), revealing that the simplified model is a poorer fit to the data.
    - Additionally, the simplified model has a larger MSE and RMSE than the original model. The MSE of the simplified model is 0.71, which is greater than the original model's MSE of 0.56. This means that the simplified model contains a larger degree of error between predicted and actual values, resulting in reduced accuracy compared to the original model. Furthermore, the simplified model has a RMSE of 0.84, which is greater than the original model's RMSE of 0.75. This reveals that the simplified model's predictions deviate from the actual median house value by $84,000 on average, which is $9,000 more than the original model. Greater average deviation indicates that the simplified model is less accurate and precise than the original model.
- Because the simplified model produces greater deviation than the original model and can account for less of the variation between its predictions and actual values, the simplified model performs poorly compared to the full model. This demonstrates that there is a tradeoff between model simplicity and performance. Basing a model on fewer features may enhance the model's explainability, but it sacrifices a degree of accuracy by excluding features that may significantly influence the target variable.
    
Would you use this simplified model in practice? Why or why not?
- I would not use this simplified model in practice. Although it bolsters the model's explainability and efficiency, the simplified model produces less accurate predictions for median house value. By limiting the features used in model training, essential information is left out, which results in lower predictive accuracy. If there were significantly more features in the dataset, I understand the appeal of a simplified model trained only on the most influential features; however, since this dataset only includes eight features, I think it is feasible to train the model on all eight to optimize model performance without sacrificing explainability.